crates/sncast/src/starknet_commands/verify/voyager.rs

```diff
@@ -0,0 +1,361 @@
+}
+
+fn metadata_error<S>(name: S, path: S) -> anyhow::Error
+where
+    S: fmt::Display,
+{
+    anyhow!("Scarb metadata failed for {name}: {path}")
+}
+
+fn gather_packages(metadata: &Metadata, packages: &mut Vec<PackageMetadata>) -> Result<()> {
+    let mut workspace_packages: Vec<PackageMetadata> = metadata
+        .packages
+        .clone()
+        .into_iter()
+        .filter(|package_meta| metadata.workspace.members.contains(&package_meta.id))
+        .filter(|package_meta| !packages.contains(package_meta))
+        .collect();
+
+    let workspace_packages_names = workspace_packages
+        .iter()
+        .map(|package| package.name.clone())
+        .collect_vec();
+
+    // find all dependencies listed by path
+    let mut dependencies: HashMap<String, PathBuf> = HashMap::new();
+    for package in &workspace_packages {
+        for dependency in &package.dependencies {
+            let name = &dependency.name;
+            let url = Url::parse(&dependency.source.repr)
+                .map_err(|_| dependency_path(name.clone(), dependency.source.repr.clone()))?;
+
+            if url.scheme().starts_with("path") {
+                let path = url
+                    .to_file_path()
+                    .map_err(|()| dependency_path(name.clone(), dependency.source.repr.clone()))?;
+                dependencies.insert(name.clone(), path);
+            }
+        }
+    }
+
+    packages.append(&mut workspace_packages);
+
+    // filter out dependencies already covered by workspace
+    let out_of_workspace_dependencies: HashMap<&String, &PathBuf> = dependencies
+        .iter()
+        .filter(|&(k, _)| !workspace_packages_names.contains(k))
+        .collect();
+
+    for (name, manifest) in out_of_workspace_dependencies {
+        let new_meta = MetadataCommand::new()
+            .json()
+            .manifest_path(manifest)
+            .exec()
+            .map_err(|_| metadata_error(name.clone(), manifest.to_string_lossy().to_string()))?;
+        gather_packages(&new_meta, packages)?;
+    }
+
+    Ok(())
+}
+
+fn package_sources(package_metadata: &PackageMetadata) -> Result<Vec<Utf8PathBuf>> {
+    let mut sources: Vec<Utf8PathBuf> = WalkDir::new(package_metadata.root.clone())
+        .into_iter()
+        .filter_map(std::result::Result::ok)
+        .filter(|f| f.file_type().is_file())
+        .filter(|f| {
+            if let Some(ext) = f.path().extension() {
+                if ext != OsStr::new(CAIRO_EXT) {
+                    return false;
+                }
+                // Skip test files
+                let path_str = f.path().to_string_lossy();
+                if path_str.contains("tests") {
+                    return false;
+                }
+                // Skip files containing #[test] attributes
+                if let Ok(contents) = fs::read_to_string(f.path()) {
+                    if contents.contains("#[test]") {
+                        return false;
+                    }
```

Comment:
This will skip a potential file containing both contract and test - I don't think we should do that. 

cc @kkawula @cptartur